src/cmd/input.go

```diff
@@ -30,34 +34,108 @@ func readInputConfig() {
 	viper.ReadInConfig()
 }
 
-func readResourceInput[T any]() (*T, error) {
+func readInput() ([]byte, error) {
 	var err error
-	var resource T
-	var yamlData []byte
+	var input []byte
 
 	switch dataFile {
 	case ".":
-		yamlData, err = os.ReadFile("./data.yaml")
+		input, err = os.ReadFile("./data.yaml")
 	case "-":
 		if isStdInFromTerminal() {
 			log.Info().Msg("Reading input directly from command line... Press CTRL+D to stop typing")
 		}
 		buf := bytes.Buffer{}
 		_, err = buf.ReadFrom(os.Stdin)
-		yamlData = buf.Bytes()
+		input = buf.Bytes()
 	default:
-		yamlData, err = os.ReadFile(dataFile)
+		input, err = os.ReadFile(dataFile)
 	}
 	if err != nil {
-		return nil, err
+		return input, err
 	}
+	return input, nil
+}
 
-	if err := yaml.Unmarshal(yamlData, &resource); err != nil {
+func ReadResource[T any](input []byte) (*T, error) {
+	var resource T
+	if err := yaml.Unmarshal(input, &resource); err != nil {
 		return nil, err
 	}
 	return &resource, nil
 }
 
+func handleJSONSchema(m map[string]any) error {
+	// since a YAML object is map[string]any which is compatible with JSONSchema so nothing needs to be done.
+	// if schema is a string, the JSON inside the string needs to be parsed. supports multiline.
+	if _, ok := m["schema"].(string); !ok {
+		return nil
+	}
+	jsonSchema, err := opslevel.NewJSONSchema(m["schema"].(string))
+	if err != nil {
+		return err
+	}
+	m["schema"] = jsonSchema
+	return nil
+}
+
+func handleJSONString(m map[string]any) error {
+	// TODO: multiline JSON is not supported as a value input because handling whitespace and escape characters is nontrivial
+	jsonString, err := opslevel.NewJSONInput(m["value"])
+	if err != nil {
+		return err
+	}
+	m["value"] = jsonString
+	return nil
+}
+
+func ReadResourceHandleJSONFields[T any](input []byte) (*T, error) {
+	var err error
+	if input == nil {
+		input, err = readInput()
+		if err != nil {
+			return nil, fmt.Errorf("error reading from input: %w", err)
+		}
+	}
+
+	m, err := ReadResource[map[string]any](input)
+	if err != nil {
+		return nil, fmt.Errorf("error creating map from input: %w", err)
+	}
+	toMap := *m
+
+	if _, ok := toMap["schema"]; ok {
+		err := handleJSONSchema(toMap)
+		if err != nil {
+			return nil, fmt.Errorf("error from handleJSONSchema: %w", err)
+		}
+	}
+	if _, ok := toMap["value"]; ok {
+		err := handleJSONString(toMap)
+		if err != nil {
+			return nil, fmt.Errorf("error from handleJSONString: %w", err)
+		}
+	}
+
+	var finalInput T
+	err = mapstructure.Decode(toMap, &finalInput)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding map as type %T: %w", finalInput, err)
+	}
+	return &finalInput, nil
+}
+
+func ReadResourceInput[T any](input []byte) (*T, error) {
```

Comment:
The only reason `input` is here is because of the tests - we should refactor this so it doesn't take input (because the only thing i should be doing is "reading input"

The function that takes an input argument should be a helper function in the test code.

Reply:
@rocktavious being able to pass in `input` can serve a purpose outside of tests as well, for example by being to do `readInput()` to get input in bytes that can be used to get the raw input from the user before calling `ReadResourceInput()`.

Reply:
This add complexity for a "what if" that we are currently not doing in our code outside of testing nor would likely do IMO.

Is there any reason that the tests cannot just use `ReadResource[T](input)` ?

Reply:
@rocktavious I have changed the `ReadResourceInput` signature to use a variadic argument for the mocked inputs. Is that OK? It means that we don't need to `(nil)` every time.

Reply:
thats definitely better - i still don't like it because now you are abusing variadic - especially because you  don't handle multiple arguments - but we can move on since this gets what i'm after it just leaves tech debt in our codebase.